main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
 object Inliner extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
 
   /**
-    * The maximum score of a function for it to be eligible for inlining.
+    * Performs inlining on the given AST `root`.
     */
-  val MaxScore = 25
+  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+    implicit val genSym: GenSym = flix.genSym
+
+    val definitions = root.definitions
+
+    val scores = definitions.foldLeft(Map.empty[DefnSym, Int]) {
+      case (macc, (sym, constant)) =>
+        macc + (sym -> Score.score(constant))
+    }
+
+    val inlinedDefinitions =
+    definitions.foldLeft(Map.empty[DefnSym, Constant]) {
+      case (macc, (sym,constant)) =>
+        val newExp = inlineExpr(definitions, scores, constant.exp)
+      macc + (sym -> constant.copy(exp = newExp))
+    }
+
+    root.copy(definitions = inlinedDefinitions).toSuccess
+
+  }
 
   /**
-    * Performs inlining on the given AST `root`.
+    * Performs the variable substitutions from `varSyms` into `exp`
     */
-  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+  def substituteWith(varSyms: Map[VarSym, VarSym], exp: Expression): Expression =
```

Comment:
In Flix, we always prefix `VarSym` with `Symbol`, so we write it as `Symbol.VarSym`.